src/transforms/compound.rs

```diff
@@ -31,82 +24,131 @@ impl GenerateConfig for CompoundConfig {
     }
 }
 
+impl CompoundConfig {
+    fn consistent_types(&self) -> bool {
+        let mut pairs = self.steps.windows(2).map(|items| match items {
+            [a, b] => (a.output_type(), b.input_type()),
+            _ => unreachable!(),
+        });
+
+        !pairs.any(|pair| {
+            matches!(
+                pair,
+                (DataType::Log, DataType::Metric) | (DataType::Metric, DataType::Log)
+            )
+        })
+    }
+}
+
 #[async_trait::async_trait]
 #[typetag::serde(name = "compound")]
 impl TransformConfig for CompoundConfig {
-    async fn build(&self, _context: &TransformContext) -> crate::Result<Transform> {
-        Err("this transform must be expanded".into())
-    }
-
-    fn expand(
-        &mut self,
-    ) -> crate::Result<Option<(IndexMap<String, Box<dyn TransformConfig>>, ExpandType)>> {
-        let mut map: IndexMap<String, Box<dyn TransformConfig>> = IndexMap::new();
-        for (i, step) in self.steps.iter().enumerate() {
-            if map
-                .insert(
-                    step.id.as_ref().cloned().unwrap_or_else(|| i.to_string()),
-                    step.transform.to_owned(),
-                )
-                .is_some()
-            {
-                return Err("conflicting id found while expanding transform".into());
-            }
-        }
-
-        if !map.is_empty() {
-            Ok(Some((map, ExpandType::Serial { alias: false })))
+    async fn build(&self, context: &TransformContext) -> crate::Result<Transform> {
+        if !self.consistent_types() {
+            Err("Inconsistent type in a compound transform".into())
         } else {
-            Err("must specify at least one transform".into())
+            Compound::new(self.clone(), context)
+                .await
+                .map(Transform::task)
         }
     }
 
     fn input_type(&self) -> DataType {
-        DataType::Any
+        self.steps
+            .first()
+            .map(|t| t.input_type())
+            .unwrap_or(DataType::Any)
```

Comment:
It looks like we require there to be at least one transform in `steps`, so we could assert that `first` returns `Some` instead of using an `Any` fallback that should never be needed.